### PR TITLE
Extract select_hidden_states and sample from model to reduce warmup from O(m*n) to O(m+n)

### PR DESCRIPTION
**Pull request by py4**
# Description

Reduces warmup time from O(m*n) to O(m+n). This is also what ptxla does.

buildkite passed: https://buildkite.com/tpu-commons/tpu-commons-ci/builds/523

warmup time for llama:
* after: 2 min
* before: 16 min

# Tests

on both vllm and jax model impl i get:

```
-------------------------------------------------
Prompt: 'Hello, my name is'
Generated text: ' Emily and I am a 25-year-old freelance writer and editor. I have'
--------------------------------------------------
Prompt: 'The capital of France is'
Generated text: ' a city of romance, art, fashion, and cuisine. Paris is a must'
--------------------------------------------------
Prompt: 'The colors of the rainbow are'
Generated text: ' often remembered using the acronym ROYGBIV, which stands for Red, Orange'
--------------------------------------------------
Prompt: 'The future of AI is'
Generated text: ' bright, but it also raises concerns about bias, accountability, and the impact on'
--------------------------------------------------
Prompt: 'The president of the United States is'
Generated text: ' the head of state and head of government of the United States. The president serves'
--------------------------------------------------
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

**Comment by wwl2755-google** [comment]
Just curious, would there be any performance degradation on splitting the jitted kernel on forward pass from one to multiple? (Like one fused cudagraph vs multiple small cudagraphs on GPU) vLLM adopted piece-wise cudagraph because of flexibility. How would we prioritize it, if it sacrifices performance?

**Comment by py4** [comment]
Yeah that’s a great point and I also have thought about it. One big graph
gives XLA flexibility to compile better. I’m planning to benchmark this to
understand the trade off and see how things are

**Comment by py4** [comment]
I profiled both codes. 

Before (decode to decode is ~15.9 ms):

image

After (decode to decode is ~15.9ms):

image

The conclusion is i don't observe a performance degradation. @hfan 's intuition was that sampling has data dependency on logits, so there is no much room for "smart" planning/fusing of XLA.
